unstructured/ingest/connector/pinecone.py

```diff
@@ -0,0 +1,125 @@
+import json
+import multiprocessing as mp
+import typing as t
+from dataclasses import dataclass
+
+from unstructured.ingest.error import DestinationConnectionError, WriteError
+from unstructured.ingest.interfaces import (
+    BaseConnectorConfig,
+    BaseDestinationConnector,
+    BaseIngestDoc,
+    BaseSessionHandle,
+    ConfigSessionHandleMixin,
+    WriteConfig,
+    WriteConfigSessionHandleMixin,
+)
+from unstructured.ingest.logger import logger
+from unstructured.utils import requires_dependencies
+
+
+@dataclass
+class PineconeSessionHandle(BaseSessionHandle):
+    service: "pinecone.Index"  # noqa: F821
+
+
+@dataclass
+class PineconeWriteConfig(WriteConfigSessionHandleMixin, ConfigSessionHandleMixin, WriteConfig):
+    api_key: str
+    index_name: str
+    environment: str
+    # todo: fix buggy session handle implementation
+    # with the bug, session handle gets created for each batch,
+    # rather than with each process
+
+    @requires_dependencies(["pinecone"], extras="pinecone")
+    def create_pinecone_object(self, api_key, index_name, environment):
+        import pinecone
+
+        pinecone.init(api_key=api_key, environment=environment)
+        index = pinecone.Index(index_name)
+        logger.debug(f"Connected to index: {pinecone.describe_index(index_name)}")
+        return index
+
+    def create_session_handle(self) -> PineconeSessionHandle:
+        service = self.create_pinecone_object(self.api_key, self.index_name, self.environment)
+        return PineconeSessionHandle(service=service)
+
+    @requires_dependencies(["pinecone"], extras="pinecone")
+    def upsert_batch(self, batch):
+        import pinecone.core.client.exceptions
+
+        index = self.session_handle.service
+        try:
+            response = index.upsert(batch)
+        except pinecone.core.client.exceptions.ApiException as api_error:
+            raise WriteError(f"http error: {api_error}") from api_error
+        logger.debug(f"results: {response}")
+
+
+@dataclass
+class SimplePineconeConfig(BaseConnectorConfig):
+    api_key: str
+    index_name: str
+    environment: str
+
+
+@dataclass
+class PineconeDestinationConnector(BaseDestinationConnector):
+    write_config: WriteConfig
+    connector_config: SimplePineconeConfig
+
+    @DestinationConnectionError.wrap
+    @requires_dependencies(["pinecone"], extras="pinecone")
+    def initialize(self):
+        pass
+
+    def write_dict(self, *args, dict_list: t.List[t.Dict[str, t.Any]], **kwargs) -> None:
+        logger.info(
+            f"Inserting / updating {len(dict_list)} documents to destination "
+            f"index at {self.connector_config.index_name}",
+        )
+
+        # this is advised to be 100 at maximum in pinecone docs, however when we
+        # chunk content, we hit to the object size limits, so we decrease the batch
+        # size even more here
+        pinecone_batch_size = 10
```

Comment:
does this need to scale with chunk size?

Reply:
Has an inverse relationship with anything that increases the amount of data in an element, including chunk_size

So technically we could break this if we had huge documents with huge chunk size params

But even before breaking this part, chunk size might break embeddings, since they might have fixed context windows (not sure, provider organization might have handled it as well).

Reply:
sorry, yea, get the inverse relationship, just meant that the fixed value seems suspicious. do you know the limits at which this falls apart? as long as we can handle reasonable windows, probably good enough for now and we can stub a ticket for more investigation later

Reply:
## Some research on the limits

1. For each vector, the maximum metadata size allowed is 40,960 bytes [1]. This per-vector limitation makes chunks bigger than 40,000 characters fail when doing upserts. While that is the case, 38,000 character chunks had successful upserts. (when using the example file book-war-and-peace-1225p.txt) _So the maximum eligible chunk_size was between 38,000 and 40,000_ for the embedding model that I was using (langchain-huggingface). When the embedding model is changed to have higher vector dimensions, we might come up to another limitation ie per-vector total object size limitation (regardless of metadata).

2. As for the allowed message length, the allowed maximum is 4,194,304 bytes [2]. That makes ~ x100 for the maximum vector metadata size allowed. Also for the request size, the allowed size is 2MBs [3]. These two combined limitations _allowed for 85 sized batches while not allowing for 90_. That is a bit lower than the recommended batch size (100) in [pinecone docs](https://docs.pinecone.io/reference/upsert).

> [1]: HTTP response body: {"code":3,"message":"Metadata size is ... bytes, which exceeds the limit of 40960 bytes per vector","details":[]}
> [2]: HTTP response body: {"code":11,"message":"Error, message length too large: found ... bytes, the limit is: 4194304 bytes","details":[]}
> [3]: HTTP response body: {"code":3,"message":"Request size ... exceeds the maximum supported size of 2MB","details":[]}

## Resulting Changes

I think allowing `--batch-size` as a cli param still makes sense, so I've added it with [make batch size a parameter](https://github.com/Unstructured-IO/unstructured/pull/1774/commits/766d4853daa708a052a4aa7bc1f6e433da49c3df) and I've set the default to 50.

So now we have batch size as a cli param, with the default value set to 50, and at least with `langchain-huggingface` embedder and `book-war-and-peace-1225p.txt` , it can support up to 85 as a value.